pkg/resourcemanager/controller/health/health_checker.go

```diff
@@ -29,74 +32,93 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// CheckHealth checks whether the given `runtime.Unstructured` is healthy.
-// `nil` is returned when the `runtime.Unstructured` has kind which is not supported by this function.
-func CheckHealth(ctx context.Context, c client.Client, scheme *runtime.Scheme, obj runtime.Object) error {
-	switch obj.GetObjectKind().GroupVersionKind().GroupKind() {
+// CheckHealth checks whether the given object is healthy.
+// It returns a bool indicating whether the object was actually checked and an error if any health check failed.
+func CheckHealth(ctx context.Context, c client.Client, scheme *runtime.Scheme, obj runtime.Object) (bool, error) {
+	// We must not rely on TypeMeta to be set in objects as decoder clears apiVersion and kind fields, see
+	// https://github.com/kubernetes/kubernetes/issues/80609 and https://github.com/gardener/gardener/issues/5357#issuecomment-1040150204.
+	// Instead of using GetObjectKind(), we use the scheme to figure out the GroupVersionKind which works for both typed
+	// and unstructured objects.
+	gvk, err := apiutil.GVKForObject(obj, scheme)
+	if err != nil {
+		if runtime.IsNotRegisteredError(err) {
+			// types that this function runs health checks on must be registered in the given scheme
+			// if the scheme doesn't recognize it the object, skip the check as we probably don't have a corresponding health check for it
+			// TODO: probably we should have a dedicated scheme for this func in order to prevent regressions when changing the
+			// target client scheme.
```

Comment:
The problem I'm having is: the function basically relies on the given scheme for doing its job. If a scheme is passed without the needed types in it, the function will silently skip the health checks.
This means, we can't be sure that the function correctly works in unit tests, because we are depending on the scheme that is plumbed through in the controller.
A dedicated scheme for this func would allow us to verify functionality in an encapsulated manner.

However, with a dedicated scheme we would need to add types to it, every time we want to cover more objects with health checks. But I think, this should be fine.

WDYT @rfranzke?

Reply:
> However, with a dedicated scheme we would need to add types to it, every time we want to cover more objects with health checks. But I think, this should be fine.

When we have the unit/integration tests then we would probably add a new test case when we cover more objects with health checks, I assume/expect. Consequently, if somebody misses adding the respective type to this dedicated scheme then the test would fail, right? In this case, I think it's fine to go with this approach.

Reply:
Yes, exactly. Ok, then I will go for it :)